solidity/contracts/token-distribution/TokenDistributor.sol

```diff
@@ -0,0 +1,201 @@
+    // TODO:Update docs
+    function claim(
+        address _recipient,
+        address _destination,
+        uint8 _v,
+        bytes32 _r,
+        bytes32 _s,
+        uint256 _index,
+        uint256 _amount,
+        bytes32[] calldata _merkleProof
+    ) external {
+        require(_recipient != address(0), "recipient address cannot be zero");
+        require(
+            _destination != address(0),
+            "destination address cannot be zero"
+        );
+        require(merkleRoot != "", "tokens were not allocated yet");
+        require(!isClaimed(_index), "tokens already claimed");
+
+        // Verify the signature over destination address.
+        require(
+            _recipient ==
+                recoverSignerAddress(
+                    abi.encodePacked(_destination),
+                    _v,
+                    _r,
+                    _s
+                ),
+            "invalid signature of destination address"
+        );
+
+        // Verify the merkle proof.
+        bytes32 node = keccak256(abi.encodePacked(_index, _recipient, _amount));
+        require(
+            MerkleProof.verify(_merkleProof, merkleRoot, node),
+            "invalid proof"
+        );
+
+        // Mark it claimed and send the token.
+        setClaimed(_index);
+
+        token.safeTransfer(_destination, _amount);
+
+        emit TokensClaimed(_index, _recipient, _destination, _amount);
+    }
+
+    /// Allocates amount of tokens for the merkle root.
+    /// @param _merkleRoot The merkle root.
+    /// @param _amount The amount of tokens allocated for the merkle root.
+    /// @param _unclaimedUnlockDuration Duration of a period after which unclaimed
+    /// tokens can be recovered from the contract. If the value is zero the
+    /// recovery won't be allowed.
+    function allocate(
+        bytes32 _merkleRoot,
+        uint256 _amount,
+        uint256 _unclaimedUnlockDuration
+    ) public onlyOwner {
+        require(merkleRoot == "", "tokens were already allocated");
+        require(_merkleRoot != "", "merkle root cannot be empty");
+
+        token.safeTransferFrom(msg.sender, address(this), _amount);
+
+        merkleRoot = _merkleRoot;
+
+        // If unclaimed unlock duration was provided calculate timestamp after
+        // which unclaimed tokens will be recoverable. If the duration is set to
+        // zero the tokens won't be recoverable.
+        if (_unclaimedUnlockDuration > 0) {
+            unclaimedUnlockTimestamp =
+                /* solium-disable-next-line security/no-block-members */
+                block.timestamp +
+                _unclaimedUnlockDuration;
+        }
+
+        emit TokensAllocated(_merkleRoot, _amount);
+    }
+
+    // Tokens not claimed within a given timeout should go to a treasury
+    // wallet address set on that contract.
+    // TODO: Is it fine to provide address on function call or do we want to declare
+    // it upfront on token deployment or allocation?
```

Comment:
I think it is better to provide an address as a parameter to this function. More flexible and less gas-expensive.